Makefile

```diff
@@ -224,55 +247,68 @@ endif
+# 3rd-level make.
+build_core: $(CONFIGURATION_PREFIX).bin
 
 # This middleware target is needed because the parent make does not know the
 # name to give to the release library yet.
 # $(API_RELEASE) is defined using platform-, configuration-, and
 # build-mode-specific definitions, so this target can only be called by the
 # 3rd-level make.
-api_release: $(API_RELEASE)
+build_api: $(API_RELEASE)
+
+# Generate the needed folders if they do not exist.
+$(CONFIGURATION_PREFIX)/$(CORE_DIR):
+	mkdir -p $(CONFIGURATION_PREFIX)/$(CORE_DIR)
+$(CONFIGURATION_PREFIX)/$(API_DIR):
+	mkdir -p $(CONFIGURATION_PREFIX)/$(API_DIR)
+$(dir $(API_RELEASE)):
+	mkdir -p $(dir $(API_RELEASE))
 
 # Generate the pre-linked uVisor binary for the given platform, configuration
 # and build mode.
 # The binary is then packaged with the uVisor library implementations into a
 # unique .a file.
-$(API_RELEASE): $(API_ASM_OUTPUT) $(API_DIR)/lib/$(PLATFORM)/$(BUILD_MODE) $(API_OBJS)
+$(API_RELEASE): $(dir $(API_RELEASE)) $(CONFIGURATION_PREFIX)/$(API_DIR) $(API_OBJS)
 	$(AR) Dcr $(API_RELEASE) $(API_OBJS)
 	echo "$(PROGRAM_VERSION)" > $(API_VERSION)
 
-# Generate the output asm file from the asm input file and the INCBIN'ed binary.
-$(API_ASM_OUTPUT): $(CONFIGURATION_PREFIX).bin
-	cp -f $(API_ASM_HEADER) $(API_ASM_OUTPUT)
-	$(CPP) -w -P $(API_CONFIG) $(API_ASM_INPUT) > $(API_ASM_OUTPUT)
-
-# Generate the library folders if they do not exist.
-$(API_DIR)/lib/$(PLATFORM)/$(BUILD_MODE):
-	mkdir -p $(API_DIR)/lib/$(PLATFORM)/$(BUILD_MODE)
-
 # Copy the core ELF file into a binary.
 $(CONFIGURATION_PREFIX).bin: $(CONFIGURATION_PREFIX).elf
 	$(OBJCOPY) $< -O binary $@
 
-# Link all object files into the core ELF.
-$(CONFIGURATION_PREFIX).elf: $(OBJS) $(CONFIGURATION_PREFIX).linker
+# Link all the core object files into the core ELF.
+$(CONFIGURATION_PREFIX).elf: $(CONFIGURATION_PREFIX)/$(CORE_DIR) $(OBJS) $(CONFIGURATION_PREFIX).linker
 	$(CC) $(LDFLAGS) -o $@ $(OBJS) $(SYSLIBS)
 	$(OBJDUMP) -d $@ > $(CONFIGURATION_PREFIX).asm
 
 # Pre-process the core linker script.
 $(CONFIGURATION_PREFIX).linker: $(CORE_LINKER_DIR)/default.h
-	mkdir -p $(dir $(CONFIGURATION_PREFIX))
 	$(CPP) -w -P $(LINKER_CONFIG) $< -o $@
 
-# Pre-process and compile a C file into an object file.
-%.c.$(CONFIGURATION_LOWER).$(BUILD_MODE).o: %.c
+# Pre-process and compile a core C file into an object file.
+$(CONFIGURATION_PREFIX)/$(CORE_DIR)/%.o: %.c
 	$(CC) $(CPPFLAGS) $(CFLAGS) -c $< -o $@
 
-# Pre-process and compile an S file into an object file.
-%.S.$(CONFIGURATION_LOWER).$(BUILD_MODE).o: %.S.$(CONFIGURATION_LOWER).$(BUILD_MODE).s
+# Pre-process and compile a release library C file into an object file.
+$(CONFIGURATION_PREFIX)/$(API_DIR)/%.o: %.c
 	$(CC) $(CPPFLAGS) $(CFLAGS) -c $< -o $@
 
+# Generate the output asm file from the asm input file and the INCBIN'ed binary.
+$(API_ASM_OUTPUT:.s=.o): $(CONFIGURATION_PREFIX).bin $(API_ASM_INPUT)
+	cp -f $(API_ASM_HEADER) $(API_ASM_OUTPUT)
```

Comment:
@meriac Fixed! Now it's only `cp`, the `-f` was not needed. If the file is absent we get an error, which is better.